tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/AutoTuner.scala

```diff
@@ -0,0 +1,435 @@
+
+  def get_shuffle_partitions: Int = properties("spark.sql.shuffle.partitions").toInt
+
+  def get_max_partition_bytes: String = properties("spark.sql.files.maxPartitionBytes")
+
+  def get_pinned_pool_size: String = properties("spark.rapids.memory.pinnedPool.size")
+
+  def get_executor_memory_overhead: String = properties("spark.executor.memoryOverhead")
+
+  def get_all_properties: Map[String, String] = properties.toMap
+
+  def set_executor_instances(numInstances: Int): Unit = {
+    properties("spark.executor.instances") = numInstances.toString
+  }
+
+  def set_executor_cores(executorCores: Int): Unit = {
+    properties("spark.executor.cores") = executorCores.toString
+  }
+
+  def set_executor_memory(executorMemory: String): Unit = {
+    properties("spark.executor.memory") = executorMemory
+  }
+
+  def set_concurrent_gpu_tasks(concurrentGpuTasks: Int): Unit = {
+    properties("spark.rapids.sql.concurrentGpuTasks") = concurrentGpuTasks.toString
+  }
+
+  def set_task_resource_gpu(taskResourceGpu: Double): Unit = {
+    properties("spark.task.resource.gpu.amount") = taskResourceGpu.toString
+  }
+
+  def set_shuffle_partitions(shufflePartitions: Int): Unit = {
+    properties("spark.sql.shuffle.partitions") = shufflePartitions.toString
+  }
+
+  def set_max_partition_bytes(maxPartitionBytes: String): Unit = {
+    properties("spark.sql.files.maxPartitionBytes") = maxPartitionBytes
+  }
+
+  def set_pinned_pool_size(pinnedPoolSize: String): Unit = {
+    properties("spark.rapids.memory.pinnedPool.size") = pinnedPoolSize
+  }
+
+  def set_executor_memory_overhead(executorMemoryOverhead: String): Unit = {
+    properties("spark.executor.memoryOverhead") = executorMemoryOverhead
+  }
+
+  def set_executor_memory_overhead_factor(executorMemoryOverheadFactor: Double): Unit = {
+    properties("spark.executor.memoryOverheadFactor") = executorMemoryOverheadFactor.toString
+  }
+}
+
+class AutoTuner(app: ApplicationSummaryInfo, workerInfo: String) extends Logging {
+  import AutoTuner._
+  val DEFAULT_SHUFFLE_PARTITION_MULTIPLIER: Int = 2
+  val MAX_JVM_GCTIME_FRACTION: Double = 0.3
+
+  val DEFAULT_CONCURRENT_GPU_TASKS_MULTIPLIER: Double = 0.125 // Currently aggressively set to 1/8
+  val MAX_CONCURRENT_GPU_TASKS: Int = 4
+
+  val DEFAULT_MAX_PARTITION_BYTES: String = "512m"
+  val MAX_PARTITION_BYTES_BOUND: String = "4g"
+  val MAX_PARTITION_BYTES_RANGE: String = "256m"
+  val MIN_PARTITION_BYTES_RANGE: String = "128m"
+
+  val DEFAULT_PINNED_POOL_SIZE: String = "2g"
+  val DEFAULT_MEMORY_OVERHEAD_FACTOR: Double = 0.1
+  val MIN_MEMORY_OVERHEAD = "2g"
+  val MAX_EXTRA_SYSTEM_MEMORY = "2g"
+
+  val MAX_PER_EXECUTOR_CORE_COUNT: Int = 16
+  val MIN_PER_EXECUTOR_CORE_COUNT: Int = 4
+
+  val MAX_EXECUTOR_MEMORY: String = "64g"
+  val MIN_EXECUTOR_MEMORY: String = "8g"
+
+  var comments: Seq[String] = Seq()
+
+  private def recommendSparkProperties(bestConfig: Config, systemProps: SystemProps): Unit = {
+    if (systemProps == null) {
```

Comment:
it might be nice to split this up into smaller functions to make it more readable for specific things, like a function for memory, function for cores, etc.. generally that makes unit testing easier as well

Reply:
Split into smaller functions for each property.